File: Data_Structure_Design/Q2. Insert Delete GetRandom O(1) - Duplicates allowed/Q2_InsertDeleteGetRandomO(1)-Duplicatesallowed.py
import random  # import random module for getRandom function

class RandomizedCollection(object):

    def __init__(self):
        self.nums = []  # list to store values
        self.idx = {}   # dictionary to store value -> set of indices

    def insert(self, val):
        """
        :type val: int
        :rtype: bool
        """
        res = val not in self.idx  # check if val is new
        if res:
            self.idx[val] = set()  # create a new set if val not seen
        self.idx[val].add(len(self.nums))  # store current index of val
        self.nums.append(val)  # add val to list
        return res  # return True if new, else False

    def remove(self, val):
        """
        :type val: int
        :rtype: bool
        """
        if val not in self.idx or not self.idx[val]:  # if val not found
            return False  # cannot remove
        remove_idx = self.idx[val].pop()  # get index to remove
        last = self.nums[-1]  # get last value in list
        self.nums[remove_idx] = last  # move last value to removed spot
        self.idx[last].add(remove_idx)  # update index of last value
        self.idx[last].discard(len(self.nums) - 1)  # remove old index of last value
        self.nums.pop()  # remove last element
        if not self.idx[val]:  # if val no longer in list
            del self.idx[val]  # delete val from map
        return True  # successfully removed

    def getRandom(self):
        """
        :rtype: int
        """
        return random.choice(self.nums)  # return random element from list

File: Data_Structure_Design/Q2. Insert Delete GetRandom O(1) - Duplicates allowed/Q2_InsertDeleteGetRandomO(1)-Duplicatesallowed.java
import java.util.*;

class RandomizedCollection {

    // List to store all elements (including duplicates)
    private List<Integer> nums;

    // Map to store each value and all its indices in nums
    private Map<Integer, Set<Integer>> idxMap;

    // Random object for random number generation
    private Random rand;

    public RandomizedCollection() {
        nums = new ArrayList<>();          // Initialize list
        idxMap = new HashMap<>();          // Initialize map
        rand = new Random();               // Initialize random object
    }

    public boolean insert(int val) {
        // Check if val is new (not present before)
        boolean isNew = !idxMap.containsKey(val);

        // Add new index (current end of list) to map
        idxMap.computeIfAbsent(val, k -> new HashSet<>()).add(nums.size());

        // Add val to list
        nums.add(val);

        // Return true if val was not in collection before
        return isNew;
    }

    public boolean remove(int val) {
        // If val not found in collection, return false
        if (!idxMap.containsKey(val) || idxMap.get(val).isEmpty()) {
            return false;
        }

        // Get one index where val exists
        int removeIdx = idxMap.get(val).iterator().next();

        // Get last value in list
        int lastVal = nums.get(nums.size() - 1);

        // Replace the removed value with the last value
        nums.set(removeIdx, lastVal);

        // Update map: remove this index from val
        idxMap.get(val).remove(removeIdx);

        // Add new index for lastVal
        idxMap.get(lastVal).add(removeIdx);

        // Remove old index (the end one) for lastVal
        idxMap.get(lastVal).remove(nums.size() - 1);

        // Remove last element from list
        nums.remove(nums.size() - 1);

        // If val has no more indices, remove it from map
        if (idxMap.get(val).isEmpty()) {
            idxMap.remove(val);
        }

        // Return true (removal successful)
        return true;
    }

    public int getRandom() {
        // Pick a random index and return that value
        return nums.get(rand.nextInt(nums.size()));
    }
}

/**
 * Your RandomizedCollection object will be instantiated and called as such:
 * RandomizedCollection obj = new RandomizedCollection();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */
